File: sut/ailidani.paxi/launcher.py
import json
import logging
import os

from sut.abstract import Launcher
from src.utils import helper
from src.utils.dependency_probes import DEPENDENCIES

REPO = "git@github.com:ailidani/paxi.git"
COMMIT_HASH = "6823d0b0fb1690a906391bcd5b4e0b01486ea2bd"

OPTIONS = [{"num": 0, "text": "Start Paxi"},
           {"num": 1, "text": "Stop Paxi"},
           {"num": 2, "text": "Run Benchmark"}]

PROTOCOLS = [{"num": 1, "text": "paxos", "consistency": "Linearizability"},
             {"num": 2, "text": "epaxos", "consistency": "Linearizability"},
             {"num": 3, "text": "sdpaxos", "consistency": "Linearizability"},
             {"num": 4, "text": "wpaxos", "consistency": "Linearizability"},
             {"num": 5, "text": "abd", "consistency": "Linearizability"},
             {"num": 6, "text": "chain", "consistency": "Linearizability"},
             {"num": 7, "text": "vpaxos", "consistency": "Linearizability"},
             {"num": 8, "text": "wankeeper", "consistency": "Linearizability"},
             {"num": 9, "text": "kpaxos", "consistency": "Linearizability"},
             {"num": 10, "text": "paxos_groups", "consistency": "Linearizability"},
             {"num": 11, "text": "dynamo", "consistency": "Eventual"},
             {"num": 12, "text": "blockchain", "consistency": "Linearizability"},
             {"num": 13, "text": "m2paxos", "consistency": "Linearizability"},
             {"num": 14, "text": "hpaxos", "consistency": "Linearizability"}]


class PaxiLauncher(Launcher):
    def __init__(self, nodes, ssh, client_ip, num_of_nodes, output_file):
        logging.info("Creating instance of PaxiLauncher")
        super().__init__(nodes, ssh, client_ip, num_of_nodes, output_file)

    def map_ip_port(self):
        port = {"private": {}, "public": {}}
        for node in self.nodes:
            public_ip = node["public_ip"]
            private_ip = node["private_ip"]

            if (public_ip not in port["public"]
                    or port["public"][public_ip] is None):
                port["public"][public_ip] = 3000
            else:
                port["public"][public_ip] += 1

            if (private_ip not in port["private"]
                    or port["private"][private_ip] is None):
                port["private"][private_ip] = 2000
            else:
                port["private"][private_ip] += 1

        return port

    def launch(self):
        logging.info("Launching PaxiLauncher")
        prot_num = helper.get_option(1, len(PROTOCOLS), PROTOCOLS)

        self.project_name = "ailidani.paxi"
        self.project_repository = REPO
        self.project_commit = COMMIT_HASH
        self.ycsb_interface = "paxi"
        self.ycsb_endpoint = "rest.endpoint"
        self.selected_protocol = {
            "name": PROTOCOLS[prot_num-1]["text"],
            "language": "Go",
            "consistency": PROTOCOLS[prot_num-1]["consistency"],
            "persistency": "In-Memory",
        }

        port_map = self.map_ip_port()
        self.build()

        while True:
            val = helper.get_option(0, len(OPTIONS) - 1, OPTIONS)
            print()

            match val:
                case 0:
                    self.start(port_map)
                case 1:
                    self.stop(port_map)
                case 2:
                    endpoints = [f"http://{ip}:{port}" for ip,
                                 port in port_map["public"].items()]
                    self.ycsb(endpoints)

    def generate_config(self, port_map):
        logging.info("Generating run_config.json file")
        with open(f"{self.local_dir}/template.json", 'r') as file:
            data = json.load(file)

        for i, node in enumerate(self.nodes):
            id = f"1.{i+1}"
            public_ip = node["public_ip"]
            private_ip = node["private_ip"]
            public_port = port_map['public'][public_ip]
            private_port = port_map['private'][private_ip]

            data["address"][id] = f"tcp://{private_ip}:{private_port}"
            data["http_address"][id] = f"http://{public_ip}:{public_port}"

        config = f"{self.local_dir}/run_config.json"
        with open(config, "w") as f:
            json.dump(data, f, indent=2)

        return config

    def start(self, port_map):
        config_path = self.generate_config(port_map)
        binary = f"{self.repo_dir_path}/server"
        source_files = f"{config_path} {binary}"
        remote_dir = f"/home/{self.user}/{self.project_name}"

        # Send binary to remote machine
        for node in self.nodes:
            if node["public_ip"] == "127.0.0.1":
                continue

            logging.info(f"Sending protocol executables to {node["public_ip"]}")
            self.remote_rsync(node["public_ip"], source_files, remote_dir)

        # Start paxi instances
        for i, node in enumerate(self.nodes):
            logging.info(f"Starting Paxi instance on {node["public_ip"]}")
            if node["private_ip"] == "127.0.0.1" and node["public_ip"] == "127.0.0.1":
                run_cmd = (
                    f"nohup {binary} -id 1.{i+1} "
                    f"-algorithm={self.selected_protocol['name']} "
                    f"-config {config_path} > /dev/null 2>&1 &"
                )
                self.local_run_cmd(run_cmd)
            else:
                remote_dir = f"/home/{self.user}/{self.project_name}"
                remote_binary = f"{remote_dir}/server"
                remote_config = f"{remote_dir}/run_config.json"
                run_cmd = (
                    f"nohup {remote_binary} -id 1.{i+1} "
                    f"-algorithm={self.selected_protocol['name']} "
                    f"-config {remote_config} > /dev/null 2>&1 &"
                )
                self.remote_run_cmd(node["public_ip"], run_cmd, True)

        logging.info(f"All paxi {self.selected_protocol['name']} instances successfully started")

    def stop(self, port_map):
        binary = f"{self.repo_dir_path}/server"

        for i, node in enumerate(self.nodes):
            logging.info(f"Stopping Paxi instance on {node["public_ip"]}")
            if node["private_ip"] == "127.0.0.1" and node["public_ip"] == "127.0.0.1":
                stop_cmd = (
                    f"pids=$(ps aux | grep '{binary}' | grep -v grep | awk '{{print $2}}'); "
                    f"for pid in $pids; do echo \"Killing $pid\"; kill -9 $pid; done; "
                    f"rm server.*.log; "
                )
                self.local_run_cmd(stop_cmd)
            else:
                remote_dir = f"/home/{self.user}/{self.project_name}"
                remote_binary = f"{remote_dir}/server"
                remote_config = f"{remote_dir}/run_config.json"
                stop_cmd = (
                    f"pids=$(ps aux | grep '{remote_binary}' | grep -v grep | awk '{{print $2}}'); "
                    f"for pid in $pids; do echo \"Killing $pid\"; kill -9 $pid; done; "
                    f"rm /home/{self.user}/server.*.log; "
                    f"rm {remote_config}"
                )

                self.remote_run_cmd(node["public_ip"], stop_cmd, False)

        local_config = f"{self.local_dir}/run_config.json"
        rm_config_cmd = f"rm {local_config}"
        self.local_run_cmd(rm_config_cmd)

        logging.info(f"All paxi {self.selected_protocol['name']} instances successfully stopped")

    def build(self):
        logging.info("Checking if protocol executables already exists...")
        path, matching_commit = self.ensure_repo_exists(self.local_dir,
                                                        self.project_repository,
                                                        COMMIT_HASH)
        self.repo_dir_path = path
        self.check_dependency(DEPENDENCIES["golang"], ">=1.18")
        binary = f"{path}/bin/server"

        # Rebuild if binary doesn't exist
        # or if repo commit doesn't match the default commit hash
        if not os.path.isfile(binary) or not matching_commit:
            logging.info("Building protocol executables...")
            build_cmd = (
                f"cd {path}/bin && "
                "./build.sh"
            )
            self.local_run_cmd(build_cmd)

        logging.info("Paxi build complete")

File: sut/ailidani.paxi/launcher.go
package paxi

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/fadhilkurnia/distro/internal/config"
	"github.com/fadhilkurnia/distro/internal/gitrepo"
	"github.com/fadhilkurnia/distro/internal/k6"
	"github.com/fadhilkurnia/distro/internal/launcher"
	"github.com/fadhilkurnia/distro/internal/nix"
	"github.com/fadhilkurnia/distro/internal/registry"
	"github.com/fadhilkurnia/distro/internal/runner"
)

var meta = launcher.ProjectMeta{
	Name:    "ailidani.paxi",
	WorkDir: "sut/ailidani.paxi",
	RepoURL: "https://github.com/ailidani/paxi.git",
}

// Paxi's curated list of benchmarkable revisions
var Versions = []launcher.Version{
	{Name: "master", CommitHash: "6823d0b0fb1690a906391bcd5b4e0b01486ea2bd"},
	// Add more entries here, ex:
	// {Name: "post-optimization", CommitHash: "<full 40-char hash>"},
}

// Paxi's full catalog of supported algorithms
var Specs = []launcher.Specification{
	{Protocol: "paxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "epaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "sdpaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "wpaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "abd", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "chain", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "vpaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "wankeeper", 	Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "kpaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "paxos_groups", 	Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "dynamo", 		Language: "Go", Consistency: "Eventual", 	Persistency: "In-Memory"},
	{Protocol: "blockchain", 	Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "m2paxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
	{Protocol: "hpaxos", 		Language: "Go", Consistency: "Linearizability", Persistency: "In-Memory"},
}

func init() {
	registry.AddProject(registry.Project{
		Name:       meta.Name,
		Repository: meta.RepoURL,
		NewLauncher: func(spec launcher.Specification, version launcher.Version) launcher.Launcher {
			return &PaxiLauncher{ProjectMeta: meta, spec: spec, version: version}
		},
		Specifications: Specs,
		Versions:       Versions,
	})
}

type PaxiLauncher struct {
	launcher.ProjectMeta // gives ProjectName(), l.WorkDir, l.RepoURL
	spec                 launcher.Specification
	version              launcher.Version
	addresses            []launcher.NodeAddress
}

func (l *PaxiLauncher) Specification() launcher.Specification { return l.spec }
func (l *PaxiLauncher) Version() launcher.Version             { return l.version }
func (l *PaxiLauncher) Addresses() []launcher.NodeAddress     { return l.addresses }

func (l *PaxiLauncher) Build(ctx context.Context, pool *runner.Pool, nodes []config.Node, sshCfg config.SSHConfig, progress launcher.Progress) error {
	replicas := config.ReplicaNodes(nodes)
	short := gitrepo.ShortHash(l.version.CommitHash)
	repoDir := filepath.Join(l.WorkDir, "repo")

	progress.Debug("(If not exists) Cloning %s repository...", l.RepoURL)
	if err := gitrepo.EnsureCloned(repoDir, l.RepoURL); err != nil {
		progress.Error(err)
		return err
	}

	progress.Debug("Checking out to %s", l.version.CommitHash)
	if err := gitrepo.Checkout(repoDir, l.version.CommitHash); err != nil {
		progress.Error(err)
		return err
	}

	localBinPath := filepath.Join(l.WorkDir, ".build", short, "bin", "server")
	if !gitrepo.FileExists(localBinPath) {
		local := runner.NewLocalRunner(l.WorkDir)
		script := gitrepo.ResolveOverride(l.WorkDir, "scripts/build.sh", l.version.Name)
		env := map[string]string{"HASH": short}

		progress.Info("Executing %s to compile binaries...", script)
		if err := nix.Run(ctx, local, script, env); err != nil {
			errMsg := fmt.Errorf("build failed: %w", err)
			progress.Error(errMsg)
			return errMsg
		}

		if !gitrepo.FileExists(localBinPath) {
			errMsg := fmt.Errorf("%s completed but %s was not produced", script, localBinPath)
			progress.Error(errMsg)
			return errMsg
		}
	} else {
		progress.Info("%s already built, skipping compile", localBinPath)
	}
 
	relBinPath := fmt.Sprintf(".build/%s/bin/server", short)
	for _, n := range replicas {
		if err := ctx.Err(); err != nil { return err }

		r, err := launcher.GetRunner(pool, n, l.ProjectMeta, progress)
		if err != nil { return err }

		progress.Info("Sending %s to %s (%s): %s", localBinPath, n.ID, n.PublicIP, relBinPath)
		if err := r.SendToNode(ctx, localBinPath, relBinPath); err != nil {
			errMsg := fmt.Errorf("Copying binary to %s: %w", n.ID, err)
			progress.Error(errMsg)
			return errMsg
		}
	}
	return nil
}

func computePortMap(nodes []config.Node) []launcher.NodeAddress {
	nextPublicPort := map[string]int{}
	nextPrivatePort := map[string]int{}
	out := make([]launcher.NodeAddress, len(nodes))
 
	for i, n := range nodes {
		pubPort, ok := nextPublicPort[n.PublicIP]
		if !ok {
			pubPort = 3000
		}
		nextPublicPort[n.PublicIP] = pubPort + 1
 
		privPort, ok := nextPrivatePort[n.PrivateIP]
		if !ok {
			privPort = 2000
		}
		nextPrivatePort[n.PrivateIP] = privPort + 1
 
		out[i] = launcher.NodeAddress{
			NodeID:      n.ID,
			PublicIP:    n.PublicIP,
			PrivateIP:   n.PrivateIP,
			PublicPort:  pubPort,
			PrivatePort: privPort,
		}
	}
	return out
}

func buildConfigJSON(templatePath string, addresses []launcher.NodeAddress) ([]byte, error) {
	raw, err := os.ReadFile(templatePath)
	if err != nil {
		return nil, fmt.Errorf("reading template %s: %w", templatePath, err)
	}
 
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, fmt.Errorf("parsing template %s: %w", templatePath, err)
	}
 
	address := make(map[string]string, len(addresses))
	httpAddress := make(map[string]string, len(addresses))
	for i, a := range addresses {
		id := fmt.Sprintf("1.%d", i+1)
		address[id] = fmt.Sprintf("tcp://%s:%d", a.PrivateIP, a.PrivatePort)
		httpAddress[id] = fmt.Sprintf("http://%s:%d", a.PrivateIP, a.PublicPort)
	}
	data["address"] = address
	data["http_address"] = httpAddress
 
	out, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encoding generated config: %w", err)
	}
	return out, nil
}

func (l *PaxiLauncher) Start(ctx context.Context, pool *runner.Pool, nodes []config.Node, progress launcher.Progress) error {
	replicas := config.ReplicaNodes(nodes)
	short := gitrepo.ShortHash(l.version.CommitHash)
 
	templateRel := gitrepo.ResolveOverride(l.WorkDir, "template.json", l.version.Name)
	templateAbs := filepath.Join(l.WorkDir, templateRel)
 
	addresses := computePortMap(nodes)
	progress.Debug("Generating config from %s", templateRel)
	configJSON, err := buildConfigJSON(templateAbs, addresses)
	if err != nil {
		progress.Error(err)
		return err
	}
 
	localConfigPath := filepath.Join(l.WorkDir, ".build", short, "run_config.json")
	progress.Debug("Writing generated config as %s", localConfigPath)
	if err := os.WriteFile(localConfigPath, configJSON, 0644); err != nil {
		errMsg := fmt.Errorf("Writing %s: %w", localConfigPath, err)
		progress.Error(errMsg)
		return errMsg
	}
 
	relConfigPath := fmt.Sprintf(".build/%s/run_config.json", short)
	relBinPath := fmt.Sprintf(".build/%s/bin/server", short)
	script := gitrepo.ResolveOverride(l.WorkDir, "scripts/start.sh", l.version.Name)
 
	for i, n := range replicas {
		r, err := launcher.GetRunner(pool, n, l.ProjectMeta, progress)
		if err != nil { return err }

		progress.Debug("Check if %s exists in %s (%s)", relBinPath, n.ID, n.PublicIP)
		if err := launcher.CheckBinaryExists(ctx, r, relBinPath); err != nil {
			progress.Error(err)
			return err
		}

		progress.Debug("Sending %s to %s (%s): %s", localConfigPath, n.ID, n.PublicIP, relConfigPath)
		if err := r.SendToNode(ctx, localConfigPath, relConfigPath); err != nil {
			errMsg := fmt.Errorf("Copying config to %s: %w", n.ID, err)
			progress.Error(errMsg)
			return errMsg
		}

		env := map[string]string{
			"HASH":      short,
			"NODE_ID":   fmt.Sprintf("1.%d", i+1),
			"ALGORITHM": l.spec.Protocol, // Paxi's own -algorithm flag name
		}

		progress.Info("Executing %s to start protocol in %s (%s)...", script, n.ID, n.PublicIP)
		if err := nix.Run(ctx, r, script, env); err != nil {
			errMsg := fmt.Errorf("starting on %s: %w", n.ID, err)
			progress.Error(errMsg)
			return errMsg
		}
	}
 
	l.addresses = addresses
	return nil
}

func (l *PaxiLauncher) Stop(ctx context.Context, pool *runner.Pool, nodes []config.Node, progress launcher.Progress) error {
	replicas := config.ReplicaNodes(nodes)
	short := gitrepo.ShortHash(l.version.CommitHash)
	script := gitrepo.ResolveOverride(l.WorkDir, "scripts/stop.sh", l.version.Name)
 
	for _, n := range replicas {
		r, err := launcher.GetRunner(pool, n, l.ProjectMeta, progress)
		if err != nil { return err }
 
		env := map[string]string{"HASH": short}

		progress.Info("Stopping protocol in %s (%s)...", n.ID, n.PublicIP)
		if err := nix.Run(ctx, r, script, env); err != nil {
			errMsg := fmt.Errorf("Stopping on %s: %w", n.ID, err)
			progress.Error(errMsg)
			return errMsg
		}
	}
	return nil
}

// Paxi does not support adding a new peer to a running cluster. Its
// config is loaded once at startup with no way to change membership
// while it runs.
func (l *PaxiLauncher) SupportsAddNewPeer() bool { return false }

func (l *PaxiLauncher) SupportsMultiClientMode() bool { return true }

func (l *PaxiLauncher) AddNewPeer(ctx context.Context, pool *runner.Pool, nodes []config.Node, newPeer config.Node, progress launcher.Progress) (time.Time, time.Time, error) {
	return time.Time{}, time.Time{}, fmt.Errorf("%s does not support adding a new peer", meta.Name)
}

func (l *PaxiLauncher) AwaitDataPlaneReady(ctx context.Context, target config.Node, pollInterval time.Duration, progress launcher.Progress) (time.Time, error) {
	return time.Time{}, fmt.Errorf("%s does not support adding a new peer", meta.Name)
}

func (l *PaxiLauncher) RunAddNewPeerBenchmark(ctx context.Context, pool *runner.Pool, nodes []config.Node, newPeer config.Node, params launcher.AddNewPeerParams, progress launcher.Progress) (string, error) {
	return "", fmt.Errorf("%s does not support adding a new peer", meta.Name)
}
 
func (l *PaxiLauncher) Clean(ctx context.Context, pool *runner.Pool, nodes []config.Node, removeRepo bool, progress launcher.Progress) error {
	replicas := config.ReplicaNodes(nodes)
	short := gitrepo.ShortHash(l.version.CommitHash)
	relBinPath := fmt.Sprintf(".build/%s/bin/server", short)
	relVersionDir := fmt.Sprintf(".build/%s", short)
 
	for _, n := range replicas {
		r, err := launcher.GetRunner(pool, n, l.ProjectMeta, progress)
		if err != nil { return err }
 
		progress.Debug("Checking if %s is still running in %s (%s)", relBinPath, n.ID, n.PublicIP)
		checkCmd := fmt.Sprintf(`! (ps aux | grep %s | grep -v grep > /dev/null)`, relBinPath)
		if err := r.Run(ctx, checkCmd, nil); err != nil {
			errMsg := fmt.Errorf("process still running on %s, stop it first", n.ID)
			progress.Error(errMsg)
			return errMsg
		}
	}
 
	for _, n := range replicas {
		r, err := launcher.GetRunner(pool, n, l.ProjectMeta, progress)
		if err != nil { return err }
 
		progress.Info("Cleaning %s (%s)...", n.ID, n.PublicIP)
		if err := r.Run(ctx, fmt.Sprintf("rm -rf %s", relVersionDir), nil); err != nil {
			errMsg := fmt.Errorf("cleaning %s on %s: %w", relVersionDir, n.ID, err)
			progress.Error(errMsg)
			return errMsg
		}
	}
 
	localVersionDir := filepath.Join(l.WorkDir, ".build", short)
	progress.Info("Removing all local %s", localVersionDir)
	if err := os.RemoveAll(localVersionDir); err != nil {
		errMsg := fmt.Errorf("removing local %s: %w", localVersionDir, err)
		progress.Error(errMsg)
		return errMsg
	}
 
	if removeRepo {
		repoDir := filepath.Join(l.WorkDir, "repo")
		progress.Info("Removing %s repo...", repoDir)
		if err := os.RemoveAll(repoDir); err != nil {
			errMsg := fmt.Errorf("removing %s: %w", repoDir, err)
			progress.Error(errMsg)
			return errMsg
		}
	}
 
	return nil
}

// Generates a k6 script from scripts/latency.js plus a scenarios block 
// built from Addresses() via the shared k6 package, sends it to client, runs k6
// there via scripts/run-latency.sh, and fetches the resulting summary
// JSON back to this Version's .benchmarks/latency directory
func (l *PaxiLauncher) RunLatencyBenchmark(ctx context.Context, pool *runner.Pool, nodes []config.Node, params launcher.LatencyParams, progress launcher.Progress) (string, error) {
	client := config.ClientNode(nodes)
	if len(l.addresses) == 0 {
		errMsg := fmt.Errorf("no addresses recorded, run Start first")
		progress.Error(errMsg)
		return "", errMsg
	}
	short := gitrepo.ShortHash(l.version.CommitHash)

	scriptRel := gitrepo.ResolveOverride(l.WorkDir, "scripts/latency.js", l.version.Name)
	scriptAbs := filepath.Join(l.WorkDir, scriptRel)

	genLocalPath, requestInterval, err := k6.GenerateLatencyScript(l.WorkDir, short, scriptAbs, l.addresses, params.RequestWorkload)
	if err != nil {
		progress.Error(err)
		return "", err
	}

	r, err := launcher.GetRunner(pool, client, l.ProjectMeta, progress)
	if err != nil { return "", err }

	runScript := gitrepo.ResolveOverride(l.WorkDir, "scripts/run-latency.sh", l.version.Name)
	return k6.RunAndFetchLatencyResult(ctx, r, genLocalPath, runScript, l.WorkDir, short, params, requestInterval, l.spec, progress)
}

// compile-time check that PaxiLauncher satisfies Launcher
var _ launcher.Launcher = (*PaxiLauncher)(nil)
